README.md
[![Build Status](https://travis-ci.org/cmanaha/angular-elasticsearch-logger.svg)](https://travis-ci.org/cmanaha/angular-loggly-logger)
[![Coverage Status](https://coveralls.io/repos/cmanaha/angular-elasticsearch-logger/badge.svg?branch=master&service=github)](https://coveralls.io/github/cmanaha/angular-elasticsearch-logger?branch=master)
[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/cmanaha/angular-elasticsearch-logger?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)

# AngularJS Elasticsearch Logger Service

This module provides an implementation of an AngularJS Service (actually a provider implementation) that allows user to log into [Elasticsearch](http://elastic.io)

### Installation

To install the module use:

```
bower install angular-elasticsearch-logger --save
```

### Getting started

There are a few settings that need to be setup for the module to work properly.
Make sure that your angular application defines a dependency to the module

```javascript
angular
	.module('yourAngularApp', [
		...
    	'cmanaha.angular-elasticsearch-logger'
  	]);
```

Then your module/application needs a section for the Provider tha configures the _CMRESLoggerProvider_ such as the one as follows:

```javascript
angular
    .module('yourAngularApp')
    .config(['CMRESLoggerProvider',function ( esLoggingProvider) {
        esLoggingProvider.setElasticSearchConfig({
            'host': 'http://localhost:9200',
            'apiVersion': '1.7'
        });

       esLoggingProvider.setLogConfig({
            'index': 'demo_app_index',
            'type': 'jslog',
            'bufferSize': 2500,
            'flushIntervalInMS': 3000
        });

        esLoggingProvider.setApplicationLogContext({
            'appNameTag': 'MyDemoApp',
            'envTag': 'Development'
        });
    }]);
```
In this case the configuration points to a cluster at http://localhost:9200. the _esLoggingProvider.setElasticSearchConfig(config)_ takes a configuration object with any value defined accepted by [elasticsearch.js library](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/configuration.html)

_esLoggingProvider.setLogConfig(config)_ takes an object that can be used to set up the details of how the logger will connect and insert into Elasticsearch such as the name of the index where logs will be stored.

The _esLoggingProvider.setApplicationLogContext(config)_ takes an object with any arbitrary keys and values, this gets appended to any new log line that gets inserted to elasticsearch as an extra field to the document; Applications may add extra parameters to identify and create categories in their application (for example to differenciate between environments).

**Note:** you need to add to your scripts dependencies the following scripts, 
```html
<script src="bower_components/elasticsearch/elasticsearch.js"></script>
<script src="bower_components/elasticsearch/elasticsearch.angular.js"></script>
<script src="bower_components/stacktrace-js/stacktrace.js"></script>
<script src="bower_components/angular-elasticsearch-logger/angular-elasticsearch-logger.js"></script>
```

The last step required is to make sure you add the dependency with your services/controllers and use the _CMRESLogger_ service.

```javascript
angular
    .module('yourAngularApp')
    .controller( 'MyCtrl', ['CMRESLogger',function(CMRESLogger){
		CMRESLogger.info( 'hello world' );
		CMRESLogger.debug( 'hello world' );
		CMRESLogger.warning( 'hello world' );
		CMRESLogger.error( 'hello world' );
    }]);
```

The logger provide some functionality similar to log4net and log4j where aside from passing the log message the user can pass a thread context map; This can be used to 
as instrumentation and can be quite useful to log additional information that can be 
charted with Kibana. 

The following example shows an operation where aside from the log, the duration of an
operation is stored into elasticsearch as additional contextual parameter so it can
be graphed and instrumented not only just as a log message.

```javascript
angular
    .module('yourAngularApp')
    .controller( 'MyCtrl', ['CMRESLogger',function(CMRESLogger){
        ...
        //perform operation and get instrumentation timings

        var timeInMS = 440;

        var instrumentationContext = {};
        extraInstrumentationCtx['totalTimeInMS'] =  timeInMS;
        extraInstrumentationCtx['operationType'] = 'http_op';

        CMRESLogger.info( 'Completed operation in '+timeInMS, extraInstrumentationCtx);

    }]);
```



### Configuration

The _CMRESLoggerProvider_ does also accept an extra configuration object that configures the internal behaviour of the logger on things such as how to buffer and submit to elasticsearch.

The service internally buffers all the logs and sends the documents using the [bulk API](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html#api-bulk) to ensure the scalability of log insertion.

 - _index_ : (string) Is the index that will be used (additionally the date will be post-fixed)
 - _type_ : (string) Elasticsearch documet type that should be used when inserting documents
 - _bufferSize_ : (integer) The number of log to buffer before submitting a bulk request to Elasticsearch. If that number if exceeded the logs are flushed
 - _flushIntervalInMS_ : (integer) number of milliseconds before inserting documents that might be pending from insertion in the buffer

An example of such configuration is:
```javascript
angular
    .module('yourAngularApp')
    .config(['CMRESLoggerProvider',function ( esLoggingProvider) {
        esLoggingProvider.setElasticSearchConfig({
            'host': 'http://localhost:9200',
            'apiVersion': '1.7'
        });

       esLoggingProvider.setLogConfig({
            'index': 'demo_app_index',
            'type': 'jslog',
            'bufferSize': 2500,
            'flushIntervalInMS': 3000
        });

        esLoggingProvider.setApplicationLogContext({
            'appNameTag': 'MyDemoApp',
            'envTag': 'Development'
        });
    }]);
```

### Demo

The [DEMO](demo) folder provides a simple applicaiton and configuration that can be used to test the library. For the demo to work, just point your browser to the index.html file and start elasticsearch in your development server

*NOTE*: If your elasticsearch server runs in a remote server, before executing the demo,  amend the [app.js](demo/app.js) and point the host element to your Elasticsearch configuration

Kibana3
![alt tag](images/kibana3_demo1.png)
![alt tag](images/kibana3_demo2.png)
![alt tag](images/kibana3_demo3.png)
![alt tag](images/kibana3_demo4.png)
![alt tag](images/kibana4_demo1.png)
![alt tag](images/kibana4_demo2.png)

## Dependencies
 - required:
	1. elasticsearch.js
	2. stacktrace.js

## Known issues / TODO
 - _stacktrace_ : Currently stacktrace is using an old version of the product. It really need to be revamped (and will be done in new versions)
 - _$log decoration_ : It is feasible to decorate AngularJS $log service with this elasticsearch service... still working on it and some issues with cyclical dependencies; The development branch has currently an example of such implementation however there are still cyclic dependencies problems on it.


#License
This module is licensed by Carlos Manzanedo Rueda with MIT License

The MIT License (MIT)

Copyright (c) <2015> <Carlos Manzanedo Rueda>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.



CHANGELOG.md
Version numbers correspond to `bower.json` version

# 0.0.2

## Features
Added support for context in implementation

## Bug Fixes

## Breaking Changes

test/karma.conf.js
// to debug tests run: 
//	node_modules/karma/bin/karma start ./test/karma.conf.js --browsers=Chrome --single-run=false --debug
// from the repo root directory, then use chrome for a javascript 
// debugging session

module.exports = function(config){
    var sourcePreprocessors = 'coverage';
    
    function isDebug(argument) {
      return argument === '--debug';
    };
    
    if (process.argv.some(isDebug)) {
      sourcePreprocessors = [];
    }

    config.set({

        basePath : '../',

        files : [
            'bower_components/angular/angular.js',
            'bower_components/angular-mocks/angular-mocks.js',
            'bower_components/elasticsearch/elasticsearch.js',
            'bower_components/elasticsearch/elasticsearch.angular.js',
            'bower_components/stacktrace-js/stacktrace.js',
            'angular-elasticsearch-logger.js',
            'test/unit/**/*.js'
        ],

        customLaunchers: {
          Chrome_travis_ci: {
            base: 'Chrome',
            flags: ['--no-sandbox']
          }
        },

        autoWatch : true,

        frameworks: ['jasmine'],

        browsers : ['Chrome'],

        reporters: ['spec','coverage', 'coveralls'],

        preprocessors: {
            // source files, that you wanna generate coverage for
            // do not include tests or libraries
            // (these files will be instrumented by Istanbul)
            //'*.js': ['coverage']
            '*.js': sourcePreprocessors
        },

        plugins : [
            'karma-chrome-launcher',
            'karma-phantomjs-launcher',
            'karma-spec-reporter',
            'karma-jasmine',
            'karma-coverage',
            'karma-coveralls'
        ],

        junitReporter : {
            outputFile: 'build/reports/test-results/unit.xml',
            suite: 'unit'
        },

        coverageReporter: {
            dir : 'build/reports/coverage/',
            reporters: [
              // reporters not supporting the `file` property
              { type: 'html', subdir: 'report-html' },
              { type: 'lcov', subdir: 'report-lcov' },
              // reporters supporting the `file` property, use `subdir` to directly
              // output them in the `dir` directory
              { type: 'cobertura', subdir: '.', file: 'cobertura.txt' },
              { type: 'lcovonly', subdir: '.', file: 'report-lcovonly.txt' },
              { type: 'teamcity', subdir: '.', file: 'teamcity.txt' },
              { type: 'text', subdir: '.', file: 'text.txt' },
              { type: 'text-summary', subdir: '.', file: 'text-summary.txt' },
            ]
        }
    });
};


test/unit/elasticsearchLoggerSpec.js
'use strict';

/* jasmine specs for services go here */

describe('AngularJS ElasticSearch Logger Module:', function() {
  var logProvider,
      moduleTest = this,
      levels = ['DEBUG', 'INFO', 'WARN', 'ERROR'];

  beforeEach(function () {
    // Initialize the service provider
    // by injecting it to a fake module's config block
    var fakeModule = angular.module('testing.harness', ['cmanaha.angular-elasticsearch-logger'], function () {});
    fakeModule.config( function(CMRESLoggerProvider) {
      logProvider = CMRESLoggerProvider;
    });

    // Initialize test.app injector
    module('cmanaha.angular-elasticsearch-logger', 'testing.harness');

    // Kickstart the injectors previously registered
    // with calls to angular.mock.module
    inject(function() {});
  });


  describe( 'CMRESLoggerProvider', function() {

    it('create copies of the configuration on the get methods',function(){
        var config = {'host': 'http://192.168.1.41:9200'};
        logProvider.setElasticSearchConfig(config);
        var res = logProvider.getElasticSearchConfig();
        res['host'] = 'http://localhost:9200';
        var secondRes = logProvider.getElasticSearchConfig();
        expect(res).not.toEqual(secondRes);
    });

    it( 'can set the elasticsearch configuration settings', function() {
        var config = {'host': 'http://192.168.1.41:9200'};
        logProvider.setElasticSearchConfig(config);
        var res = logProvider.getElasticSearchConfig();
        expect(res).toEqual(config);
    });

    it( 'can set the logging settings', function() {
        var config = {
            'index': 'sentiment_analysis',
            'type': 'jslog',
            'bufferSize': 5000,
            'flushIntervalInMS': 2500
        };
        logProvider.setLogConfig(config);
        var res = logProvider.getLogConfig();
        expect(config).toEqual(res);
    });

    it( 'can set the application context settings', function() {
        var config = {
            'application': 'MyWebApplication',
            'AppVersion': '0.1',
            'environment': 'Development'
        };
        logProvider.setApplicationLogContext(config);
        var res = logProvider.getApplicationLogContext();
        expect(config).toEqual(res);
    });

  });

  describe( 'CMRESLogger', function() {
    var loggerSvc, $log, $httpBackend;
    var esConfig = {
      'host': 'http://192.168.1.41:9200',
      'apiVersion': '1.7'
    };

    var logConfig = {
      'index': 'unit_test',
      'type': 'jslog',
      'bufferSize': 1,
      'flushIntervalInMS': 1000
    };

    var appLogContextConfig = {
      'application': 'unit_test',
      'AppVersion': '0.1',
      'environment': 'Development'
    };

    var TestException = function(message) {
      var self = this;
      self.message = message;
      self.exceptionType = 'TestException';
    }

    beforeEach(function () {
      inject(function ($injector) {
        loggerSvc = $injector.get('CMRESLogger');
        $log = $injector.get('$log');
        $httpBackend = $injector.get('$httpBackend');
      });
      
      /*
      spyOn(loggerSvc, 'info').and.callFake(function(message) {
        $log.info('Calling loggerSvc at Info level wth :' +  message);
      });
      */
    });

    it('should be registered', function () {
      expect(loggerSvc).not.toBe(null);
    });

    it('will send a simple message to elasticsearch at all levels when properly configured', function () {
      logProvider.setElasticSearchConfig(esConfig);
      logProvider.setLogConfig(logConfig);
      logProvider.setApplicationLogContext(appLogContextConfig);
      
      loggerSvc.info("Test message , at INFO");
      loggerSvc.debug("Test message, at DEBUG");
      loggerSvc.warning("Test message, at WARNING");
      loggerSvc.error("Test message, at ERROR");
    });


    it('will send a message with extended properties to elasticsearch at all levels when properly configured', function () {
      logProvider.setElasticSearchConfig(esConfig);
      logProvider.setLogConfig(logConfig);
      logProvider.setApplicationLogContext(appLogContextConfig);
      
      var properties = {};
      properties['string_id'] = 'StringIDValue';
      properties['metric_int'] = 4;
      properties['metric_float'] = 4.4;
      properties['nested_object'] = {};
      properties['nested_object']['name'] =  'name_example';


      loggerSvc.info("Test message , at INFO", properties);
      loggerSvc.debug("Test message, at DEBUG", properties);
      loggerSvc.warning("Test message, at WARNING", properties);
      loggerSvc.error("Test message, at ERROR", properties);
    });

    it('will send a errorWithException to elasticsearch when properly configured', function () {
      logProvider.setElasticSearchConfig(esConfig);
      logProvider.setLogConfig(logConfig);
      logProvider.setApplicationLogContext(appLogContextConfig);
      
      loggerSvc.errorWithException("Error Message", new TestException("The Exception"));
    });

    it('will send a warningWithException to elasticsearch when properly configured', function () {
      logProvider.setElasticSearchConfig(esConfig);
      logProvider.setLogConfig(logConfig);
      logProvider.setApplicationLogContext(appLogContextConfig);
      
      loggerSvc.warningWithException("Error Message", new TestException("The Exception"));
    });

    it('will flush the messages when properly configured', function () {
      logProvider.setElasticSearchConfig(esConfig);
      logProvider.setLogConfig(logConfig);
      logProvider.setApplicationLogContext(appLogContextConfig);
      
      loggerSvc.info("Test message , at INFO");

      loggerSvc.flush();
    });

    it('when flush fails the callback method defined in the logConfig.internalLogFunction gets called', function () {
      logProvider.setElasticSearchConfig(esConfig);
      logProvider.setLogConfig(logConfig);
      logProvider.setApplicationLogContext(appLogContextConfig);
      
      //FIXME: this part of the test is not going through the right branch
      //still requires attention
      $httpBackend.expectPOST('/_bulk').respond(400,'');
      loggerSvc.info("Test message , at INFO");
    });

  });


});


Gruntfile.js
module.exports = function(grunt) {
    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),

        jshint: {
            options: {
                jshintrc: true
            },
            files: {
                src: ['angular-elasticsearch-logger.js']
            }
        },

        uglify: {
            options : {
                sourceMap: true
            },
            main: {
                files: { 'angular-elasticsearch-logger.min.js': ['angular-elasticsearch-logger.js'] }
            }
        },

        karma: {
          unit: {
            configFile: 'test/karma.conf.js',
            singleRun: true
          },
          //continuous integration mode: run tests once in PhantomJS browser.
          travis: {
            configFile: 'test/karma.conf.js',
            singleRun: true,
            browsers: ['Chrome_travis_ci']
          }
      },

      watch: {
          jshint: {
              files: ['angular-elasticsearch-logger.js'],
              tasks: ['jshint']
          },

          karma: {
              files: ['*.js', '!*.min.js'],
              tasks: ['karma:unit:run'] //NOTE the :run flag
          }
      },
	
      clean : {
	packages: ["bower_components","node_modules","build"]
      }

    });

    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.loadNpmTasks('grunt-contrib-watch');
    grunt.loadNpmTasks('grunt-karma');
    grunt.loadNpmTasks('grunt-contrib-clean');

    grunt.registerTask('default', ['jshint','uglify'] );
    grunt.registerTask('test', [ 'karma:travis' ] );
    grunt.registerTask('test-all', ['karma:unit'] );
};


angular-elasticsearch-logger.js
/**
 * @ngdoc function
 * @name esLogging.esLogProvider
 * @description
 * # MainCtrl
 * Controller of the clientSideApp
 */

angular.module('cmanaha.angular-elasticsearch-logger',['elasticsearch'])
    .provider('CMRESLogger', [function ()
{
    var self = this;
    self.esClientDetails = {};
    self.logDetails = {
        'index': 'defaul_angularjs_index',
        'type': 'jslog',
        'bufferSize': 5000,
        'flushIntervalInMS': 2500,
    };
    
    self.appDetails = {};

    self.setElasticSearchConfig = function (elasticSearchConfig) {
        self.esClientDetails = elasticSearchConfig;
    };

    self.getElasticSearchConfig = function(){
        var res = angular.copy(self.esClientDetails);
        return res;
    };
    
    self.setLogConfig = function (applicationLogConfig) {
        self.logDetails = applicationLogConfig;
    };

    self.getLogConfig = function(){
        var res = angular.copy(self.logDetails);
        return res;
    };

    self.setApplicationLogContext = function (applicationLogConfig) {
        self.appDetails = applicationLogConfig;
    };

    self.getApplicationLogContext = function(){
        var res = angular.copy(self.appDetails);
        return res;
    };


    //self.$get = ['$interval', 'esFactory', function ($interval, esFactory) {
    self.$get = ['$injector', function ($injector) {

        var esFactory = $injector.get('esFactory');
        var interval = $injector.get('$interval');
        self.logBuffer = [];

        var esFact = esFactory;
        self.esClient = esFact(self.esClientDetails);

        self.level = {
            DEBUG: 'DEBUG',
            INFO: 'INFO',
            WARN: 'WARNING',
            ERROR: 'ERROR'
        };


        self.numPadding = function (num, size) {
            var ret = num + '';
            while (ret.length < size) {
                ret = '0' + ret;
            }
            return ret;
        };

        self.getIndexName = function (indexPrefx) {
            var d = new Date(Date.now());
            var ret = indexPrefx + '-' + d.getUTCFullYear() + '.' + self.numPadding(d.getUTCMonth() + 1, 2) + '.' + self.numPadding(d.getUTCDate(), 2);
            return ret;
        };

        self.decodeStackTraceLine = function (line) {
            var lineTrc = line.split('.js');
            var method = lineTrc[0] + '.js';
            var lineNum = lineTrc[1].split(':')[1];
            var charNum = lineTrc[1].split(':')[2];
            return [method, lineNum, charNum];
        };

        self.getVariableWithDefault = function (variable, defaultValue){
            var res = 'Unknown';
            if (typeof variable === 'undefined'){
                res = defaultValue;
            } else {
                res = variable;
            }
            return res;
        };

        self.internalLog = function (logMsg, level, exception, properties) {

            //Add operation to the backlog
            self.logBuffer.push(
                {
                    index: {
                        _index: self.getIndexName(self.logDetails.index),
                        _type: self.logDetails.type,
                    }
                }
            );

            var data = {
                message: logMsg,
                level: level,
                timestamp: new Date(Date.now()).toISOString(),
                browserCodeName: self.getVariableWithDefault(navigator.appCodeName,'Unknown'),
                browserName: self.getVariableWithDefault(navigator.appName,'Unknown'),
                browserVersion: self.getVariableWithDefault(navigator.appVersion,'Unknown'),
                browserProduct: self.getVariableWithDefault(navigator.product,'Unknown'),
                browserPlatform: self.getVariableWithDefault(navigator.platform,'Unknown'),
                browserUserAgent: self.getVariableWithDefault(navigator.userAgent,'Unknown'),
                browserGeolocation: self.getVariableWithDefault(navigator.geolocation,'Unknown')
            };

            //add additional application data context
            angular.forEach(self.appDetails, function (value, key) {
                data[key] = value;
            }); 

            //add additional logging properties context
            if (typeof (properties) !== 'undefined') {
                angular.forEach( properties, function (value, key) {
                    data[key] = value;
                });
            } 

            //add details about the method, line number, etc.
            //if the log was thrown from an exception, log the exception details and
            //exception stacktrace
            var trace = [];
            var traceInfo = [];
            if (typeof (exception) === 'undefined') {
                trace = printStackTrace();
            } else {
                trace = printStackTrace(exception);
                var relevantStacktrace = trace.slice(5);
                var fullTraceStr = 'Exception Msg [ ' + exception.message + '] \n';
                angular.forEach(relevantStacktrace, function (traceLine) {
                    traceInfo = self.decodeStackTraceLine(traceLine);
                    fullTraceStr = fullTraceStr + '\t' + traceInfo[0] + ' LineNum [' + traceInfo[1] + '] charNum [' + traceInfo[2] + ']\n';
                });
                data.stacktrace = fullTraceStr;
            }
            traceInfo = self.decodeStackTraceLine(trace[5]);
            data.method = traceInfo[0];
            data.lineNum = traceInfo[1];
            data.charNum = traceInfo[2];

            //Add document to the backlog
            self.logBuffer.push(data);

            if (self.logBuffer.length / 2 >= self.logDetails.bufferSize) {
                self.internalFlush();
            }
        };

        self.internalFlush = function () {
            if (self.logBuffer.length > 0) {
                self.esClient.bulk({
                    body: self.logBuffer
                }, function (err, rsp) {
                    if (err) {
                        if (self.logDetails.internalLogFunction !== undefined){
                            self.logDetails.internalLogFunction('Got an error [' + err + ']while storing log : ' + angular.toJson(rsp, true));
                        }
                    }
                });
                self.logBuffer = [];
            }
        };

        self.flushInterval = interval(self.internalFlush, self.logDetails.flushIntervalInMS);



        return {
            info: function (message, properties) { self.internalLog(message, self.level.INFO, undefined, properties); },
            warning: function (message, properties) { self.internalLog(message, self.level.WARN, undefined, properties); },
            error: function (message, properties) { self.internalLog(message, self.level.ERROR, undefined, properties); },
            debug: function (message, properties) { self.internalLog(message, self.level.DEBUG, undefined, properties); },
            warningWithException: function (message, exception, properties) { self.internalLog(message, self.level.WARN, exception, undefined); },
            errorWithException: function (message, exception, properties) { self.internalLog(message, self.level.ERROR, exception, undefined); },
            flush: function(){self.internalFlush();},
        };
    }];
}]);


demo/app.js
'use strict';

angular
	.module('ElasticSearchLoggerDemoApp', [
    	'cmanaha.angular-elasticsearch-logger'
  	]);

angular
    .module('ElasticSearchLoggerDemoApp')
    .config(['CMRESLoggerProvider',function ( esLoggingProvider) {
        // Config parameters from 
        // http://www.elasticsearch.org/guide/en/elasticsearch/client/javascript-api/current/configuration.html

        esLoggingProvider.setElasticSearchConfig({
            //point the host:port to an elasticsearch instance or run
            //an elasticsearch service in your own
            'host': 'http://localhost:9200',
            'apiVersion': '1.7'
        });

        esLoggingProvider.setLogConfig({
            'index': 'demo_app_index',
            'type': 'jslog',
            'bufferSize': 1,
            'flushIntervalInMS': 1000
        });

        esLoggingProvider.setApplicationLogContext({
            'application': 'DemoApp',
            'version': '0.0.1',
            'environment': 'Development'
        });
    }]);


//Using decorator to override the angular $log functionality
angular.module('ElasticSearchLoggerDemoApp')
	.config(['$provide',function($provide){
	    $provide.decorator('$log',['$delegate','$injector',function($delegate, $injector){
        
        var eslogger = $injector.get('CMRESLogger');

        var wrapMethod = function( originalMethod, esLoggerMethod ){
            var retMethod = function(){
                var args = [].slice.call(arguments);
                originalMethod.apply(null,args);
                esLoggerMethod(args);
            };

            return retMethod;
        };


        //wrap the existing API
        $delegate.log =    wrapMethod($delegate.log, eslogger.info);
        $delegate.debug =  wrapMethod($delegate.debug, eslogger.debug);
        $delegate.info =   wrapMethod($delegate.info, eslogger.info);
        $delegate.warn =   wrapMethod($delegate.warn, eslogger.warn);
        $delegate.error =  wrapMethod($delegate.error, eslogger.error);

        return $delegate;

    }]);
}]);

//capture unhandled exceptions and log them to ElasticSearch
angular.module('ElasticSearchLoggerDemoApp')
  	.factory('$exceptionHandler', function() {
  		return function(exception, cause) {
    		exception.message += ' (caused by "' + cause + '")';
    		console.error(exception.message);
  		};
	});


angular
    .module('ElasticSearchLoggerDemoApp')
    .controller( 'DemoCtrl', ['$log',function($log){
    	var self = this;
        self.inputToken = null;
        self.message = '';

        self.TestException = function(message){
        	this.message = message;
        };


        self.logIt = function() {
            $log.info( self.message );
        };

        self.logMultipleTimes = function(){
            var numberOfLogs = Math.floor((Math.random() * 200) + 1);
            
            for (var i=0 ; i<numberOfLogs; i++){
                var level = Math.floor((Math.random() * 5) + 1);
                switch(level){
                    case 1: 
                        CMRESLogger.info( 'message ['+i+']: ' + self.message );
                        break;
                    case 2: 
                        CMRESLogger.debug( 'message ['+i+']: ' + self.message );
                        break;
                    case 3: 
                        CMRESLogger.warning( 'message ['+i+']: ' + self.message );
                        break;
                    case 4: 
                        CMRESLogger.error( 'message ['+i+']: ' + self.message );
                        break;
                    case 5: 
                        CMRESLogger.errorWithException( 'message ['+i+']: ' + self.message , new self.TestException('message'));
                }
            }
        };


    }]);

